### DISCUSSION TITLE
Help! Can I build my FYP using this ?

### DISCUSSION THREAD

**Discussion by uzair004** [vladmandic/human#138] [Q&A] [answered]
I want to use human to build a website where users enter details of other people along with their pics and Guest Users can upload an image to find similar people by their faces.

as new entries and pics will be added on the go, Will face models need to be trained on every added pic ?

How would be performance ?

How you would recommend implementing, as this is my first project ?

**Comment by vladmandic** [comment, marked as answer]
There is no "training", you only need to run detection on each uploaded photo and store resulting face descriptor (which is an array of float numbers).  

Assuming using NodeJS for backend implementation, backend would preload all required models and wait for request from frontend.
As user uploads the picture to frontend, it sends it to server which just runs face detection and extracts face descriptor.  

Then you store the image and corresponding face descriptor somewhere on the server (as a file or in a DB) and also keep that face descriptor in an array (each new descriptor gets added to array).  

Final step would to loop through the array and run face match between current descriptor and each descriptor in the array, sort results them by similarity and send top-n results back from server to frontend.

Performance-wise, running detection and extracting face descriptor is measured in miliseconds, so thats not a problem.
But to find best match you do have to run full loop to compare all faces so if number of faces grows that can become a bottleneck - we're talking about ~1sec for 1k compares.

Now, to get full performance from the backend that operation should be performed in a separate worker as NodeJS is by definition single-threaded - you don't want one user to hog entire server. You can check included demos for an example of NodeJS multi-process implementation.

**Reply by uzair004** [reply]
Thank you very much for detail answer, i really appreciate your efforts.

However why would i need array when i have descriptions in a file ? Or why i need file if i have all details in array ?

I am also planning to do comparison on Frontend by sending that array to Frontend from server because it will transfer load on each client rather then on single server. 
What do you think of this ? 
There will be more Network operations but I think it won't be that expensive performance wise

**Reply by vladmandic** [reply]
You store each new entry to file and to array. You don't want to re-read a file into array on each new entry, only on server startup.
